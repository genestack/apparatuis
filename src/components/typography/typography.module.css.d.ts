/*
 * Copyright (c) 2011-2019 Genestack Limited
 * All Rights Reserved
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF GENESTACK LIMITED
 * The copyright notice above does not evidence any
 * actual or intended publication of such source code.
 */
/** default body block variant */
export const root: string;
export const ellipsis: string;

/** intents */
export const quiet: string;
export const success: string;
export const warning: string;
export const alarm: string;

/** box */
export const inline: string;
export const paragraph: string;

/** inverted */
export const inverted: string;

/** variant */
export const header: string;
export const title: string;
export const section: string;
export const body: string;
export const caption: string;

/** condensed */
export const condensed: string;
