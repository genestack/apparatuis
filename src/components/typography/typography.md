```js
const {DarkContext} = require('../../utils/dark-context');

initialState = {
    inverted: false,
    quiet: true
};

<React.Fragment>
    <PageContent as={Paper}>
        <Typography as="h1" variant="header" box="paragraph">
            Small Airways Smoking Hackett 2012 SRP005411 (SRA)
        </Typography>
        <Typography as="p" box="paragraph">
            I&R, Homo sapiens,{' '}
            <Typography as="span" quiet box="inline">
                Tissue:
            </Typography>{' '}
            Epithelium of Bronchiole
        </Typography>
        <Typography as="h2" variant="title" box="paragraph">
            Description and Background
        </Typography>
        <Typography as="p" box="paragraph">
            Morphological changes in the small airway epithelium are the first histopathological
            manifestations of smoking-induced lung disease. Gene expression profiling using
            microarrays has permitted the identification of changes in the small airway epithelium
            of chronic cigarette smokers who have normal pulmonary function.
        </Typography>
        <Typography as="p" box="paragraph">
            The availability of high throughput, massively parallel RNA sequencing (RNA-Seq) of the
            entire transcriptome opens the possibility of a higher sensitivity and more detailed
            characterization of the response of the small airway epithelium to smoking.
        </Typography>
        <Typography as="h3" variant="section" box="paragraph">
            Epithelium of Bronchiole Basic information
        </Typography>
        <Typography as="p" box="paragraph">
            Epithelial cells were obtained by fiberoptic bronchoscopy and brushing of healthy
            smokers (n=6) and n=5 healthy nonsmokers, all with normal lung function and chest
            x-rays. RNA was extracted and used for massively parallel sequencing of PolyA selected
            transcripts using the Illumina Genome Snalyzer II.
        </Typography>

        <Typography as="p" variant="caption" box="paragraph">
            The availability of high throughput, massively parallel RNA sequencing (RNA-Seq) of the
            entire transcriptome opens the possibility of a higher sensitivity and more detailed
            characterization of the response of the small airway epithelium to smoking.
        </Typography>
    </PageContent>

    <Divider gap={4} variant="transparent" />

    <Controls style={{alignItems: 'flex-start'}} gap={6}>
        <ControlsItem>
            <Controls justify="end" as="label">
                <ControlsItem>
                    <Typography>Inverted</Typography>
                </ControlsItem>
                <ControlsItem>
                    <input
                        type="checkbox"
                        checked={state.inverted}
                        onChange={(event) => setState({inverted: event.currentTarget.checked})}
                    />
                </ControlsItem>
            </Controls>
            <Divider variant="transparent" />
            <Controls justify="end" as="label">
                <ControlsItem>
                    <Typography>Quiet</Typography>
                </ControlsItem>
                <ControlsItem>
                    <input
                        type="checkbox"
                        checked={state.quiet}
                        onChange={(event) => setState({quiet: event.currentTarget.checked})}
                    />
                </ControlsItem>
            </Controls>
        </ControlsItem>
        <ControlsItem shrink>
            <PageContent
                as={Paper}
                style={{background: state.inverted ? 'rgb(51, 51, 51)' : 'rgb(255, 255, 255)'}}
            >
                <DarkContext.Provider value={state.inverted}>
                    <Typography as="h1" variant="header" box="paragraph">
                        <Typography as="span" variant="header" box="inline" quiet={state.quiet}>
                            Quiet Header
                        </Typography>{' '}
                        Helvetica Bold{' '}
                        <Typography as="span" variant="title" box="inline" quiet={state.quiet}>
                            24/32
                        </Typography>
                    </Typography>
                    <Typography as="h2" variant="title" box="paragraph">
                        <Typography as="span" variant="title" box="inline" quiet={state.quiet}>
                            Quiet Title
                        </Typography>{' '}
                        Helvetica Bold{' '}
                        <Typography as="span" variant="title" box="inline" quiet={state.quiet}>
                            20/28
                        </Typography>
                    </Typography>
                    <Typography as="h3" variant="section" box="paragraph">
                        <Typography as="span" variant="section" box="inline" quiet={state.quiet}>
                            Quiet Section
                        </Typography>{' '}
                        Helvetica Bold
                        <Typography as="span" variant="body" box="inline" quiet={state.quiet}>
                            20/28
                        </Typography>
                    </Typography>
                    <Typography as="p" box="paragraph">
                        <Typography as="span" box="inline" quiet={state.quiet}>
                            Quiet body text
                        </Typography>{' '}
                        Helvetica Normal
                        <Typography as="span" variant="caption" box="inline" quiet={state.quiet}>
                            14/20
                        </Typography>
                    </Typography>
                    <Typography variant="caption" box="paragraph" quiet={state.quiet}>
                        Quiet caption text â€“ small body text Helvetica Normal 12/18
                    </Typography>
                    <Typography variant="caption" box="paragraph">
                        <Typography as="span" variant="caption" box="inline" quiet>
                            Tissue:
                        </Typography>{' '}
                        Epithelium of Bronchiole
                    </Typography>
                    <Divider gap={4} variant="transparent" />
                    <Typography box="paragraph">
                        A &lt;STRONG&gt; tag{' '}
                        <strong>
                            is working well with <em>texts</em> and <em>captions</em>
                        </strong>
                    </Typography>
                    <Typography box="paragraph">
                        <Typography quiet as="span" box="inline">
                            Quiet is useful for secondary information
                        </Typography>, while <em>Italic</em> is done with &lt;EM&gt; tag.
                    </Typography>
                </DarkContext.Provider>
            </PageContent>
        </ControlsItem>
    </Controls>
</React.Fragment>;
```
