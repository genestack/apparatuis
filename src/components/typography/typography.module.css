/*
 * Copyright (c) 2011-2019 Genestack Limited
 * All Rights Reserved
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF GENESTACK LIMITED
 * The copyright notice above does not evidence any
 * actual or intended publication of such source code.
 */

/** default body block variant */

.root {
    box-sizing: border-box;
    color: rgba(52, 52, 52, 1);
    font-weight: normal;
    display: block;
    font-size: 14px;
    line-height: 20px;
    font-family: "Helvetica", sans-serif;
    align-self: baseline;
}

/** accent */

.primary {
    font-weight: bold;
}

.secondary {
    color: rgba(0, 0, 0, 0.43);
}

/** box */

.inline {
    display: inline;
    vertical-align: baseline;
}

.paragraph {
    margin-bottom: 14px;
}

/** inversed */

.inversed {
    color: rgba(255, 255, 255, 1);
}

.inversed.secondary {
    color: rgba(136, 136, 136, 1);
}

/** headline variant */

.headline {
    font-size: 24px;
    line-height: 28px;
}

.headline.paragraph {
    margin-bottom: 24px;
}

/** title variant */

.title {
    font-size: 18px;
    line-height: 22px;
}

.title.paragraph {
    margin-bottom: 18px;
}

/** subtitle */

.subtitle {
    font-size: 16px;
    line-height: 22px;
}

.subtitle.paragraph {
    margin-bottom: 16px;
}

/** caption variant */

.caption {
    font-size: 12px;
    line-height: 16px;
}

.caption.paragraph {
    margin-bottom: 12px;
}
