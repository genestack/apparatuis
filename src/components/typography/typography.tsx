/*
 * Copyright (c) 2011-2019 Genestack Limited
 * All Rights Reserved
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF GENESTACK LIMITED
 * The copyright notice above does not evidence any
 * actual or intended publication of such source code.
 */
import classNames from 'classnames';
import * as React from 'react';

import {DarkContext} from '../../utils/dark-context';
import {mergeClassesProps, WithClasses} from '../../utils/styles';

import * as styles from './typography.module.css';

interface TargetProps {
    className?: string;
    children?: React.ReactNode;
}

type DefaultTargetProps = React.HTMLAttributes<HTMLElement>;

interface TypographyProps extends WithClasses<keyof typeof styles> {
    /**
     * Corresponds to main role and style of the text block
     * (font-size, font-weight, line-height etc.)
     *
     * Default: `"body"`
     */
    variant?: 'header' | 'title' | 'section' | 'body' | 'caption';
    /**
     * Makes text quiet. It is useful for secondary information
     */
    quiet?: boolean;
    /**
     * Describes how the text is presented in block model.
     *
     * Default: `block`
     *
     * `block` — the text is block by default.
     * `inline` — useful if you want to insert some word in a text;
     * `paragraph` — the text will have margins and paddings according to its `variant`.
     */
    box?: 'block' | 'inline' | 'paragraph';
    /**
     * Inverses the text colors on dark backgrounds
     *
     * Default: `"false"`
     */
    inverted?: boolean;
    /** Defines a color of the text */
    status?: 'success' | 'warning' | 'error';
    /**
     * Adds ellipsis style to element
     */
    ellipsis?: boolean;
    /**
     * You could redefine the target component by passing ReactType.
     *
     * Default: `"div"`
     *
     * @example
     * const anchor = <Typography as="a" /> // renders as anchor
     * const button = <Typography as={Button} /> // renders as Button component
     * const routerLink = (
     *   <Typography as={(props) => <RouterLink to="/" {...props} />} />
     * ); // renders as `RouterLink`
     */
    as?: React.ReactType;
}

/** Typography public properties */
export type Props<T extends TargetProps = DefaultTargetProps> = T & TypographyProps;

/**
 * Component that renders text with specific preset.
 * Its purpose is to reduce amount of custom CSS text styles.
 * It is one of the base components of whole UI Kit.
 * Will be moved to UI Kit.
 */
export function Typography<T extends TargetProps = DefaultTargetProps>(props: Props<T>) {
    return (
        <DarkContext.Consumer>
            {(darkContext) => {
                const {
                    as: Component = 'div',
                    variant = 'body',
                    quiet,
                    box = 'block',
                    inverted = darkContext,
                    classes,
                    className,
                    status,
                    ellipsis,
                    ...rest
                } = mergeClassesProps(props as Props<TargetProps>, styles);

                return (
                    <Component
                        {...rest}
                        className={classNames(className, classes.root, {
                            [classes.header]: variant === 'header',
                            [classes.title]: variant === 'title',
                            [classes.section]: variant === 'section',
                            [classes.body]: variant === 'body',
                            [classes.caption]: variant === 'caption',

                            [classes.quiet]: quiet,

                            [classes.inverted]: inverted,

                            [classes.inline]: box === 'inline',

                            [classes.paragraph]: box === 'paragraph',

                            [classes.success]: status === 'success',
                            [classes.warning]: status === 'warning',
                            [classes.error]: status === 'error',

                            [classes.ellipsis]: ellipsis
                        })}
                    />
                );
            }}
        </DarkContext.Consumer>
    );
}
