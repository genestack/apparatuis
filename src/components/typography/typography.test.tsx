/*
 * Copyright (c) 2011-2023 Genestack Limited
 * All Rights Reserved
 * THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF GENESTACK LIMITED
 * The copyright notice above does not evidence any
 * actual or intended publication of such source code.
 */
import * as React from 'react';
import '@testing-library/jest-dom';
import {fireEvent, render} from '@testing-library/react';

import {Typography} from './typography';

describe('Typography Component', () => {
    describe('by default', () => {
        const setup = () => render(<Typography data-testid="test">Test String</Typography>);

        test('should render single children', () => {
            expect(setup().queryByTestId('test')).toBeVisible();
        });

        test('should render p HTML element', () => {
            expect(setup().queryByTestId('test')).toBeInstanceOf(HTMLParagraphElement);
        });

        test('should render string children', () => {
            expect(setup().queryByTestId('test')).toHaveProperty('textContent', 'Test String');
        });
    });

    describe('when pass "button" string to `as` property', () => {
        const setup = () =>
            render(
                <Typography as="button" data-testid="test">
                    Test String
                </Typography>
            );

        test('should render the button element', () => {
            expect(setup().queryByTestId('test')).toBeInstanceOf(HTMLButtonElement);
        });

        test('should render valid children', () => {
            expect(setup().queryByTestId('test')).toHaveProperty('textContent', 'Test String');
        });
    });

    describe('when pass React.Component to `as` property', () => {
        function Paragraph(props: any) {
            return <span {...props} />;
        }

        const setup = () =>
            render(
                <Typography as={Paragraph} data-testid="test">
                    Test String
                </Typography>
            );

        test('should use the React.Component as root element', () => {
            expect(setup().queryByTestId('test')).toBeInstanceOf(HTMLSpanElement);
        });

        test('should render the React.Component children', () => {
            expect(setup().queryByTestId('test')).toHaveProperty('textContent', 'Test String');
        });
    });

    it('should spread all properties to the root element', () => {
        const handleClick = jest.fn();

        const screen = render(
            <Typography as="a" data-testid="test" onClick={handleClick}>
                Test String
            </Typography>
        );

        expect(screen.queryByTestId('test')).toBeInstanceOf(HTMLAnchorElement);
        fireEvent.click(screen.queryByTestId('test')!);
        expect(handleClick).toBeCalled();
    });

    it('should merge class name with own', () => {
        const screen = render(
            <Typography className="test-class-name" data-testid="test">
                Test String
            </Typography>
        );

        expect(screen.queryByTestId('test')).toHaveProperty(
            'className',
            expect.stringContaining('root')
        );
        expect(screen.queryByTestId('test')).toHaveProperty(
            'className',
            expect.stringContaining('body')
        );
        expect(screen.queryByTestId('test')).toHaveProperty(
            'className',
            expect.stringContaining('test-class-name')
        );
    });

    it('should merge classes names with own', () => {
        const screen = render(
            <Typography
                data-testid="test"
                variant="caption"
                classes={{
                    caption: 'test-caption'
                }}
            >
                Test String
            </Typography>
        );

        expect(screen.queryByTestId('test')).toHaveProperty(
            'className',
            expect.stringContaining('root')
        );
        expect(screen.queryByTestId('test')).toHaveProperty(
            'className',
            expect.stringContaining('caption')
        );
        expect(screen.queryByTestId('test')).toHaveProperty(
            'className',
            expect.stringContaining('test-caption')
        );
    });
});
