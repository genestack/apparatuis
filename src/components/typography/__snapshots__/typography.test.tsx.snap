// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography Component by default should match snapshot 1`] = `
<div
  className="root"
>
  Test String
</div>
`;

exports[`Typography Component when pass "button" string to \`as\` property should match snapshot 1`] = `
<button
  className="root"
>
  Test String
</button>
`;

exports[`Typography Component when pass React.Component to \`as\` property should match snapshot 1`] = `
<p
  className="root"
>
  Test String
</p>
`;
